include/triton/core/tritonrepoagent.h

```diff
@@ -0,0 +1,307 @@
+#define TRITONREPOAGENT_ISPEC
+#endif
+#else
+#if defined(_MSC_VER)
+#define TRITONREPOAGENT_DECLSPEC __declspec(dllimport)
+#define TRITONREPOAGENT_ISPEC __declspec(dllexport)
+#else
+#define TRITONREPOAGENT_DECLSPEC
+#define TRITONREPOAGENT_ISPEC
+#endif
+#endif
+
+struct TRITONREPOAGENT_Agent;
+struct TRITONREPOAGENT_AgentModel;
+
+///
+/// TRITONREPOAGENT API Version
+///
+/// The TRITONREPOAGENT API is versioned with major and minor version
+/// numbers. Any change to the API that does not impact backwards
+/// compatibility (for example, adding a non-required function)
+/// increases the minor version number. Any change that breaks
+/// backwards compatibility (for example, deleting or changing the
+/// behavior of a function) increases the major version number. A
+/// repository agent should check that the API version used to compile
+/// the agent is compatible with the API version of the Triton server
+/// that it is running in. This is typically done by code similar to
+/// the following which makes sure that the major versions are equal
+/// and that the minor version of Triton is >= the minor version used
+/// to build the agent.
+///
+///   uint32_t api_version_major, api_version_minor;
+///   TRITONREPOAGENT_ApiVersion(&api_version_major, &api_version_minor);
+///   if ((api_version_major != TRITONREPOAGENT_API_VERSION_MAJOR) ||
+///       (api_version_minor < TRITONREPOAGENT_API_VERSION_MINOR)) {
+///     return TRITONSERVER_ErrorNew(
+///       TRITONSERVER_ERROR_UNSUPPORTED,
+///       "triton repository agent API version does not support this agent");
+///   }
+///
+#define TRITONREPOAGENT_API_VERSION_MAJOR 0
+#define TRITONREPOAGENT_API_VERSION_MINOR 1
+
+/// Get the TRITONREPOAGENT API version supported by Triton. This
+/// value can be compared against the
+/// TRITONREPOAGENT_API_VERSION_MAJOR and
+/// TRITONREPOAGENT_API_VERSION_MINOR used to build the agent to
+/// ensure that Triton is compatible with the agent.
+///
+/// \param major Returns the TRITONREPOAGENT API major version supported
+/// by Triton.
+/// \param minor Returns the TRITONREPOAGENT API minor version supported
+/// by Triton.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONREPOAGENT_DECLSPEC TRITONSERVER_Error* TRITONREPOAGENT_ApiVersion(
+    uint32_t* major, uint32_t* minor);
+
+/// TRITONREPOAGENT_ArtifactType
+///
+/// The ways that the files that make up a model's repository content
+/// are communicated between Triton and the agent.
+///
+///   TRITONREPOAGENT_ARTIFACT_FILESYSTEM: The model artifacts are
+///     communicated to and from the repository agent via a locally
+///     accessible filesystem. The agent can access these files using
+///     an appropriate filesystem API.
+///
+typedef enum TRITONREPOAGENT_artifacttype_enum {
+  TRITONREPOAGENT_ARTIFACT_FILESYSTEM
+} TRITONREPOAGENT_ArtifactType;
+
+/// TRITONREPOAGENT_ActionType
+///
+/// Types of repository actions that can be handled by an agent.
```

Comment:
When there is change in lifecycle state, i.e. LOAD -> LOAD_FAIL. On LOAD_FAIL, what will the first call to `TRITONREPOAGENT_ModelRepositoryLocation()` returns? The last committed location from LOAD or the original repository?

Reply:
Another question is on model reloading, the specified agents in model config may have changed: i.e. from agents A -> B to B -> C. What will their lifecycles look like?

I think for B is straight forward, just calling `TRITONREPOAGENT_ModelAction()` with RELOAD. A is going to be directly calling its `TRITONREPOAGENT_ModelFinalize()`. And C is calling `TRITONREPOAGENT_ModelInitialize()` and then `TRITONREPOAGENT_ModelAction()` with RELOAD?

Reply:
Good point. I think we should remove RELOAD so that each AgentModel goes through init -> LOAD -> UNLOAD -> finalize. When a model is reloaded the existing agents will get UNLOAD -> finalize and the new agents (with a new AgentModel object) will get init -> LOAD.
That should make the flows much clearer and easier to implement.

Reply:
Also keep the *_COMPLETE states...